File: python/hot100/71.字符串解码.py
class Solution:
    def decodeString(self, s: str) -> str:
        if not s:
            return s
        if s[0].isalpha():
            return s[0] + self.decodeString(s[1:])

        i = s.find('[')
        balance = 1
        for j in count(i + 1):
            if s[j] == "[":
                balance += 1
            elif s[j] == "]":
                balance -= 1
                if balance == 0:
                    return self.decodeString(s[i+1 : j]) * int(s[:i]) + self.decodeString(s[j + 1:])

File: cpp/hot100/71.字符串解码.cpp
class Solution {
public:
    string decodeString(string s) {
        if(s.empty()){
            return s;
        }
        if(isalpha(s[0])){
            return s[0] + decodeString(s.substr(1));
        }
        int i = s.find('[');
        int balance = 1;
        for(int j = i + 1; ;++j){
            if(s[j] == '['){
                ++balance;
            }
            else if(s[j] == ']'){
                --balance;
                if(balance == 0){
                    int k = stoi(s.substr(0, i));
                    string t = decodeString(s.substr(i + 1, j - i - 1));
                    string res;
                    while(k--){
                        res += t;
                    }
                    return res + decodeString(s.substr(j + 1));
                }
            }
        }
    }
};